controlpanel/frontend/views/datasource.py

```diff
@@ -208,52 +239,103 @@ def get_success_url(self):
         return reverse_lazy("manage-datasource", kwargs={"pk": self.object.s3bucket.id})
 
 
-class GrantAccess(LoginRequiredMixin, PermissionRequiredMixin, CreateView):
+class RevokeIAMManagedPolicyAccess(RevokeAccess):
+    model = PolicyS3Bucket
+
+
+class GrantAccessMixin:
+    def get_form_kwargs(self):
+        return FormMixin.get_form_kwargs(self)
+
+    def get_success_url(self):
+        messages.success(self.request, "Successfully granted access")
+        return reverse_lazy(
+            "manage-datasource",
+            kwargs={"pk": self.object.s3bucket.id}
+        )
+
+    def form_valid(self, form):
+        user = self.request.user
+        bucket = get_object_or_404(S3Bucket, pk=self.kwargs['pk'])
+
+        if not user.has_perm('api.grant_s3bucket_access', bucket):
+            raise PermissionDenied()
+
+        is_admin = form.cleaned_data.get('is_admin')
+
+        if is_admin and not user.has_perm('api.add_s3bucket_admin', bucket):
+            raise PermissionDenied()
+
+        self.object = self.model.objects.create(**self.values(form))
+        return FormMixin.form_valid(self, form)
+
+
+class GrantAccess(
+    GrantAccessMixin,
+    LoginRequiredMixin,
+    PermissionRequiredMixin,
+    CreateView,
+):
     context_object_name = 'users3bucket'
     form_class = GrantAccessForm
     model = UserS3Bucket
     permission_required = 'api.create_users3bucket'
     template_name = 'datasource-access-grant.html'
 
-    def get_form_kwargs(self):
-        return FormMixin.get_form_kwargs(self)
-
     def get_context_data(self):
         context = super().get_context_data()
         bucket = get_object_or_404(S3Bucket, pk=self.kwargs['pk'])
         context['bucket'] = bucket
-        member_ids = list(bucket.users3buckets.all().select_related('user').values_list(
-            'user__auth0_id',
-            flat=True,
-        ))
+        member_ids = list(
+            bucket.users3buckets.all().select_related('user').values_list(
+                'user__auth0_id',
+                flat=True,
+            ))
```

Comment:
I'm not sure why this has been re-indented, as the line length is 88 chars, which is what Black defaults to. In any case, the final `))` should be separated.